src/network/networkd-ndisc.c

```diff
@@ -1013,6 +1018,51 @@ static int ndisc_drop_outdated(Link *link, usec_t timestamp_usec) {
         return r;
 }
 
+static int ndisc_drop_lifetime_zero_router(Link *link, sd_ndisc_router *rt) {
+        struct in6_addr router, prefix;
+        unsigned prefixlen, lifetime_sec;
+        Route *route;
+        int r;
+
+        assert(link);
+
+        r = sd_ndisc_router_get_address(rt, &router);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to get gateway address from RA: %m");
+
+        r = sd_ndisc_router_prefix_get_address(rt, &prefix);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to get prefix address: %m");
+
+        r = sd_ndisc_router_prefix_get_prefixlen(rt, &prefixlen);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to get prefix length: %m");
+
+        r = sd_ndisc_router_prefix_get_valid_lifetime(rt, &lifetime_sec);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to get prefix lifetime: %m");
+
+        SET_FOREACH(route, link->routes) {
+                if (route->source != NETWORK_CONFIG_SOURCE_NDISC)
+                        continue;
+
+                if (route->gw_family == AF_INET6 && memcmp(&route->gw.in6, &router, sizeof(route->gw.in6)) == 0) {
+                        r = route_remove_and_drop(route);
+                        if (r < 0)
+                                log_link_warning_errno(link, r, "Failed to remove outdated SLAAC route, ignoring: %m");
+                }
+
+                if (route->family == AF_INET6 && memcmp(&route->dst.in6, &prefix, sizeof(route->dst.in6) == 0) &&
+                    route->dst_prefixlen == prefixlen && lifetime_sec == 0)  {
+                        r = route_remove_and_drop(route);
+                        if (r < 0)
+                                log_link_warning_errno(link, r, "Failed to remove outdated SLAAC route, ignoring: %m");
+                }
```

Comment:
Such comparison is not necessary.
Please check `route->provider` and the router address.